### PR TITLE
[WIP] MSMQ to ASB bridge sample

### PR DESCRIPTION
**Pull request by SeanFeldman**
The sample
- Bridges MSMQ-based endpoint with ASB-based endpoint using NServiceBus.Bridge
- Send commands and publish events from MSMQ endpoint and handle those by ASB endpoint

Benefits of this sample:
- no native/low level transport knowledge is required
    - can be used to bridge various transports together
- bridge can be removed once a single transport is decided upon

### CODE REVIEW

**Comment by SeanFeldman** [comment]
Bridge v2 doesn't have `bridgeConfiguration.InterceptForawrding()`.

**Comment by SeanFeldman** [comment]
@Particular/azure-maintainers adding this sample to Azure samples collection.
@Particular/docs-maintainers please review (note that NSB v7 sample is broken [until Bridge v2 is fixed](https://github.com/Particular/docs.particular.net/pull/3358#issuecomment-337475640)).
/cc @indualagarsamy

**Comment by weralabaj** [comment]
> note that NSB v7 sample is broken until Bridge v2 is fixed

@SeanFeldman so do we need to wait for that fix? what's the plan?

File: samples/azure/azure-service-bus-msmq-bridge/sample.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+---
+title: MSMQ to Azure Service Bus Transport Bridge
+reviewed: 2017-10-17
+component: ASB
+related:
+- transports/azure-service-bus
+- nservicebus/bridge
+---
+
+
```
Comment (weralabaj) [inline samples/azure/azure-service-bus-msmq-bridge/sample.md]: I'd add a short paragraph explaining the purpose of this sample, in particular, in what situation it is useful (e.g. you mentioned POC, when transport is not decided yet).

File: samples/azure/azure-service-bus-msmq-bridge/sample.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+Covered scenarios are:
+
+ * Sending commands from MSMQ endpoint to Azure Service Bus endpoint.
+ * Publishing events from MSMQ endpoint and subscribing to those events from Azure Service Bus endpoint.
+
+
+### Bridging
+
+Endpoints are bridged using [Transport Brindge](/nservicebus/bridge/). `Bridge` project is implemented as a standalone process that runs side-by-side with the bridged endpoints, `MsmqEndpoint` and `AsbEndpoint`.
```
Comment (weralabaj) [inline samples/azure/azure-service-bus-msmq-bridge/sample.md]: typo Transport Brindge

Comment (weralabaj) [reply samples/azure/azure-service-bus-msmq-bridge/sample.md]: > bridged

Does that word actually exist?

File: samples/azure/azure-service-bus-msmq-bridge/sample.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+snippet: connect-msmq-side-of-bridge
+
+![MSMQ topology][msmq-topology]
+
+Command routing to Azure Service Bus endpoint is specified using bridge extension method:
+
+snippet: route-command-via-bridge
+
+NOTE: to access bridge extension method, project has to reference NServiceBus.Bridge.Connector NuGet package.
```
Comment (weralabaj) [inline samples/azure/azure-service-bus-msmq-bridge/sample.md]: > NServiceBus.Bridge.Connector

missing quote marks `NServiceBus.Bridge.Connector`

File: samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            });
+
+        bridgeConfiguration.AutoCreateQueues();
+        bridgeConfiguration.UseSubscriptionPersistece<InMemoryPersistence>((configuration, persistence) => { });
+
+        #endregion
+
+        #region suppress-transaction-scope-for-asb
+
+        bridgeConfiguration.InterceptForawrding(async (queue, message, forward) =>
```
[the remark this answers is not in the document]
Comment (SzymonPobiega) [reply samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs]: This is a typo in NSB.Bridge 1.1. Unfortunately I won't be able to fix it for 1.1. I fixed it in 2.0 (targeting NSB 7) though. Thanks ❤️

**Comment by SzymonPobiega** [comment]
This is blocked on me releasing NSB.Bridge 2 which is blocked on ASB checking the existence of a serializer: https://particularsoftware.slack.com/archives/C1607JM6G/p1508326487000203

File: samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        var bridgeConfiguration = Bridge
+            .Between<MsmqTransport>("Bridge-MSMQ")
+            .And<AzureServiceBusTransport>("Bridge-ASB", transport =>
+            {
+                transport.ConnectionString(connectionString);
+                transport.UseForwardingTopology();
+            });
+
+        bridgeConfiguration.AutoCreateQueues();
+        bridgeConfiguration.UseSubscriptionPersistece<InMemoryPersistence>((configuration, persistence) => { });
```
[the remark this answers is not in the document]
Comment (SzymonPobiega) [reply samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs]: This is a typo in NSB.Bridge 1.1. Unfortunately I won't be able to fix it for 1.1. I fixed it in 2.0 (targeting NSB 7) though. Thanks ❤️

**Comment by weralabaj** [comment]
Hmm... so let's mark as WIP until we get v7 to work?

File: samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SeanFeldman) [reply samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs]: Issue raised https://github.com/SzymonPobiega/NServiceBus.Bridge/issues/11

File: samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SeanFeldman) [reply samples/azure/azure-service-bus-msmq-bridge/ASB_7/Bridge/Program.cs]: Issue raised https://github.com/SzymonPobiega/NServiceBus.Bridge/issues/11

File: samples/azure/azure-service-bus-msmq-bridge/sample.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+
+## Code walk-through
+
+This sample shows an integration between two endpoints running on two different transports, `MsmqEndpoint` endpoint running on MSMQ and `AsbEndpoint` running on Azure Service Bus.
+
+Covered scenarios are:
+
+ * Sending commands from MSMQ endpoint to Azure Service Bus endpoint.
+ * Publishing events from MSMQ endpoint and subscribing to those events from Azure Service Bus endpoint.
```
[the remark this answers is not in the document]
Comment (SeanFeldman) [reply samples/azure/azure-service-bus-msmq-bridge/sample.md]: The sample was based on a customer need. It also helped to find some bugs and fix those.
Now that those bugs are addressed, I could send a command from one transport and raise an event from another.

**Comment by SeanFeldman** [comment]
> so do we need to wait for that fix? what's the plan?

@weralabaj Indeed. I've provided the customer with a link to this PR, so they can move forward.

File: samples/azure/azure-service-bus-msmq-bridge/sample.md (hunk 2 above)
Comment (SeanFeldman) [reply samples/azure/azure-service-bus-msmq-bridge/sample.md]: As a verb, yes.

File: samples/azure/azure-service-bus-msmq-bridge/sample.md (hunk 1 above)
Comment (SeanFeldman) [reply samples/azure/azure-service-bus-msmq-bridge/sample.md]: @weralabaj added an opening paragraph to elaborate. Have a look. Thanks.

**Comment by SeanFeldman** [comment]
@yvesgoeleven I've changed the sample to send a command from ASB, handled by MSMQ and raising an event that would be handled by ASB. Is this better?

**Comment by SeanFeldman** [comment]
@weralabaj could you please review?

**Comment by SeanFeldman** [comment]
Splitting core v7 sample into a separate PR until issues with Core/transport is resolved.

**Comment by SeanFeldman** [comment]
ASB v8/Core v7 sample: PR #3370

**Comment by indualagarsamy** [comment]
Hey @SeanFeldman - I could be wrong but it looked like a build fail on the last commit. Was the issue resolved before the merge?

**Comment by SeanFeldman** [comment]
The nuget version syntax was off. I had "1*" rather than "1.*". That caused the build to fail. With commit https://github.com/Particular/docs.particular.net/pull/3358/commits/ca094642137efa38d45fad3bb5c3eb29c9024634 it was resorted.

**Pull request merged**